---
title: "Examples"
author: Maxime Jan
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  md_document:
    variant: markdown_github
    toc: true
    toc_depth: 2
---

# Sleep-Wake Driven Models, an R package [SWDMr]

*An R package to fit models for sleep-wake driven phenotypes* 

- [x] Read and format vigilant state data
- [ ] Read *smo* file format
- [x] Fit a Process-S dynamic model
- [ ] Fit a Process-C dynamic model
- [x] Fit an driven damped harmonic oscillator model


```{r}
library(SWDMr) # Package for model construction, objective function building

# External package
library(optimx) # Package for parameter optimization
library(ggplot2) # Package for visualization
```


# Format vigilant state

Get vigilant state of mice. Each row is a 4 seconds epoch containing the vigilant state of the mouse

Wake = w | 1
NREM sleep = n | 2
REM sleep = r | 3

```{r}
data(SleepWakeData)
head(as.data.frame(SleepWakeData))
```

Compute the amount of wake,nrem and rem sleep per 6 minutes bin (0.1h)

```{r}
# Read the matrix and transform it into a data.frame for time spent in each state per "concattimesec" bin [in sec]
SWdf<-Read_SW(SleepWakeData,concattimesec = 360) # 300 = 5min, 180 = 3 min, 360 = 6min

# We can add day and night cycle
SWdf<-SWdf_AddLD(SWdf)

# We merge day 1 and 2 that are "baseline", then following days are 3 and 4, then baseline sleep is reiterated
SWdf<-SWdf_DayMerging(SWdf,Daysformat=list(c(1,2),c(1,2),3,4,c(1,2)),concattimesec=360)

# We add sleep deprivation even at time 48
SWdf<-SWdf_AddSD(SWdf,c(48,54))

# End result:
head(SWdf)
```

Sleep amount in the experiment

```{r,fig.width=6,fig.height=3}
gg<-ggplot(aes(x=Time,y=Sleep),data=SWdf)+geom_line(color="darkblue",size=1)
gg<-gg+scale_x_continuous(breaks=seq(0,96,by=12))
gg<-gg+ylab("Sleep amount [h]")+xlab("Time [h]")
gg
```

# Fit a process-S dynamic on phenotype

First we load data for time spent awake or asleep

## Data:

```{r}
# Contain a data.frame of time spent asleep or awake
data(SWdf)
head(SWdf)
```

We then load phenotypes containing also time of sampling corresponding with the sleep-wake data.frame

```{r}
data(RNAExpression)
head(RNAExpression)
```

Let's model the dynamics of Homer1

## Create model

We create a swdmr object containing sleep-wake data and phenotypes

```{r}
swdmr <- SWDMr(SWdist=SWdf, Gexp=RNAExpression)
swdmr
```

We then initiate a model for a process-S dynamics

```{r}
modelS<-initProcessSmodel(swdmr,VarExp = "Homer1")
```

We replicate the baseline time (time0-time24) 10x to have a stable baseline dynamics

```{r}
modelS<-ReplicateDrivingForce(modelS,c(0,24.0),10)
```


We fix the starting point as mean expression level

```{r}
modelS<-SetYinitMode(modelS,"Fixed",mean(RNAExpression$Homer1))
```


We want to optimize our objective function to minimize the Residual Sum of Square (RSS)

```{r}
modelS<-SetFittingValue(modelS,value = "RSS")
```

Our model as the following free parameters:

```{r}
modelS
```

*We could fix some of these parameters*

```{r}
# Here, asymptotic value for time spent awake is 10 for the phenotype of interest
anothermodel<-FixAsymptWake(modelS,10)
anothermodel
```

## Optimization

We create an objective function:

```{r}
objfun<-SWDMrGetEvalFun(modelS)
```

We use optimx to optimize the free parameter to minimize the RSS

```{r}
paramsS<-c("AsympWake"=6,AsympSleep=4,TauWake=10,TauSleep=10)
# Many other algorithms and options exist ! See optimx manual to fit your needs
fitsS<-optimx(paramsS,fn = objfun,method=c("Nelder-Mead"))
fitsS
```
## Visualize fit

Let see the fit

```{r}
outS<-SWDMrFit(modelS,fitsS)
```


```{r,fig.width=6,fig.height=3}
# plot fitted lines
gg<-ggplot(aes(x=time,y=fit),data=cbind.data.frame(time=outS$time,fit=outS$y1))
gg<-gg+geom_line()
# Add expression points
gg <- gg + annotate("point",x=RNAExpression$Time,y=RNAExpression$Homer1)
gg<-gg + xlim(-24,96)
gg
```

## Compute some statistics


```{r}
residuals<-SWDMrStats(modelS,outS,detailed = T)$residuals
fitted<-SWDMrStats(modelS,outS,detailed = T)$fitted
stats<-SWDMrStats(modelS,outS,detailed = T)$stats
```


- Residual Sum of Square: RSS
- Negative Log Likelihood: NLL
- Bayesian Information Criterion: BIC
- BIC of a linear model with an intercept only: BIC_flat
- Bayes Factor between our model and a flat model: BayesFactor
- Akaike Information Criterion: AIC
- Number of samples in model: n
- Number of free parameters: k
- Variance of resiudals: ErrorVariance
- Kendall's tau between data points and fit: KendalTau


```{r}
stats
```
*Here n=56 while dataset contains 62 points. Because fit was performed up to Time96 while last points reach Time 216 and 222. If these points needed to be integrated, then SWdf should go up to time 222*


```{r}
plot(fitted,residuals);abline(h=mean(residuals))
qqnorm(residuals);qqline(residuals)
```


# Driven Damped harmonic oscillator

$$m\frac{d^2x}{dt^2} + m\gamma\frac{dx}{dt} + m\omega^2_{0}x = F(t)$$

Build swdmr object

```{r}
## If you want a greater timestep for Rk4
# SWdf<-SWdf[rep(seq_len(nrow(SWdf)), each = 10), ]
# SWdf$Time<-seq(0.01,120.0,by=0.01)
```


```{r}
swdmr <- SWDMr(SWdist=SWdf, Gexp=RNAExpression)
swdmr
```

Initiate a Driven Damped Harmonic Oscillator [DDHO] model for a gene

```{r}
Gene<-"Arntl"
model<-initDDHOmodel(swdmr,VarExp = Gene)
```

Set some parameter of our model

```{r}
# Mean expression in baseline between highest and lowest value
MeanPerTime<-aggregate(RNAExpression[RNAExpression$Time <= 48,Gene],list(RNAExpression$Time[RNAExpression$Time <= 48]),mean)
MeanGeneExprInBaseline<-(max(MeanPerTime$x)+min(MeanPerTime$x))/2

# Fix the intercepts
model<-FixIntercept(model,MeanGeneExprInBaseline)
# Add sleep-wake force
model<-AddForce(model,"Wake")
model<-AddForce(model,"Sleep")

# Start is set at intercept with speed of 0
model<-SetYinitMode(model,mode = "Intercept_0",values = c(0,48))
# We replicate baseline for 20 day
model<-ReplicateDrivingForce(model,c(0,24.0),40)
# A sin-wave force is applied with a period of 24h
model<-AddSinF(model,FixPer = 24)
# Compute the fit using RSS
model<-SetFittingValue(model,value = "RSS")
# Penalize the fitting for unstable value for 10 replicated days
model<-PenalizeUnstableFit(model,value = T,PredictedValueInterval = c(0,48), StabilityDayCheck = 10)


```

summary of the model

```{r}
model
```

Fit data with optimx

```{r}
# Get objective function
objfun<-SWDMrGetEvalFun(model)

# 1st Fit
params<-c(Wake=-0.1,Sleep=0.1,loggamma=log(1e-1),omega=2*pi/24,AmpSin=0,PhiSin=pi)
fits<-optimx(params,fn = objfun,method=c("nlminb"),control=list(maxit=1000))

fits

if (! any(! is.na(fits$Wake))){
  params<-c(Wake=0,Sleep=0,loggamma=log(1e-1),omega=2*pi/24,AmpSin=0,PhiSin=pi)
  fits<-optimx(params,fn = objfun,method=c("newuoa"),control=list(maxit=10000))
}

if (any(fits$convcode == 0)){
  fits<-fits[fits$convcode == 0,,drop=F]
  optimxres<-fits[order(fits$value),,drop=F][1,]
}else{
  optimxres<-fits[order(fits$value),,drop=F][1,]
}

optimxres
```


Get fit

```{r}
out<-SWDMrFit(model,params = optimxres[1,])
par(mfrow=c(2,1))
plot(out$time,out$y1,type="l",ylab="Position",xlab="Time",xlim=c(0,120),xaxt="n")
axis(1,seq(0,120,by=12),seq(0,120,by=12))
plot(out$time,out$y2,type="l",xlab="Time",xlim=c(0,120),xaxt="n",ylab="RNA velocity",lwd=2)
axis(1,seq(0,120,by=12),seq(0,120,by=12))
```

Get some statistics for the fit (R2, and AdjR2 are not valid measure ! Do not use them !)

```{r}
SWDMrStats(model,out,detailed = T)$stats
```

Plot fit and gene expression using ggplot2

```{r}
SWDMr:::StandardFittingPlot(model,optimxres[1,])
```

```{r}
sessionInfo()
```

