---
title: "Examples"
author: Maxime Jan
date: 12 August 2019
output:
  md_document:
    variant: markdown_github
---

# Package

load packages

```{r}
library(SWDMr)
library(optimx)
library(ggplot2)
```

# Prepare data

Sleep-wake data per hour

```{r}
# Data
# ****** Prepare Sleep-Wake data ****** 
files <- list.files(path="../data/Sleep_States/",pattern = paste("^","BL6",sep=""),full.names = T)
SWdf<-Read_SW(files,concattimesec = 360) # 300 = 5min, 180 = 3 min, 360 = 6min
SWdf<-SWdf_AddLD(SWdf) # Add Light and Dark
SWdf<-SWdf_DayMerging(SWdf,Daysformat=list(c(1,2),c(1,2),3,4,c(1,2)),concattimesec=360)
SWdf<-SWdf_AddSD(SWdf,c(48,54))
head(SWdf)
```

Explained variable is gene expression

```{r}
# ***** Prepare Gene expression data ******
# Gene expression
load("../data/CHor2018_TimeCourse_NormalizedData_Genes.RData")
# Meta data
load("../data/CHor2018_TimeCourse_MetaData.RData")
# Remove data sampled 7days after SD
rna_expr<-rna_expr[,rna_meta$time<100]
rna_meta<-rna_meta[rna_meta$time<100,]
# Row = samples, column = gene
rna_expr<-t(rna_expr)
#rna_expr<-2^rna_expr
rna_expr<-as.data.frame(rna_expr)
# Add time factor
rna_expr$Time<-rna_meta$time+24
head(rna_expr[,c("Arntl","Dbp","Time")])
```

# Run fitting

Build swdmr object

```{r}
swdmr <- SWDMr(SWdist=SWdf, Gexp=rna_expr)
swdmr
```

Initiate a Driven Damped Harmonic Oscillator [DDHO] model for a gene

```{r}
Gene<-"Ciart"
model<-initDDHOmodel(swdmr,VarExp = Gene)
```

Set some parameter of our model

```{r}
# Mean expression in baseline between highest and lowest value
MeanPerTime<-aggregate(rna_expr[rna_meta$SD_NSD == "NSD",Gene],list(rna_meta$time[rna_meta$SD_NSD == "NSD"]),mean)
MeanGeneExprInBaseline<-(max(MeanPerTime$x)+min(MeanPerTime$x))/2

# Fix the intercepts
model<-FixIntercept(model,MeanGeneExprInBaseline)
# Add sleep-wake force
model<-AddForce(model,"Wake")
model<-AddForce(model,"Sleep")
# Start is set at intercept with speed of 0
model<-SetYinitMode(model,mode = "CircadianFit",values = c(0,48))
# We replicate baseline for 20 day
model<-ReplicateDrivingForce(model,c(0.1,24.0),20)
# A sin-wave force is applied with a period of 24h
model<-AddSinF(model,FixPer = 24)
# Compute the fit using RSS
model<-SetFittingValue(model,value = "RSS")
# Penalize the fitting for unstable value for 10 replicated days
model<-PenalizeUnstableFit(model,value = T,PredictedValueInterval = c(0,48), StabilityDayCheck = 10)
```

summary of the model

```{r}
model
```

Fit data with optimx

```{r}
# Get objective function
objfun<-SWDMrGetEvalFun(model)
# Limits of the model
params<-c(Wake=0,Sleep=0,loggamma=log(1e-1),omega=2*pi/23.8,AmpSin=1e-3,PhiSin=5.2)
lower<-c(Wake=-Inf,Sleep=-Inf,loggamma=-Inf,omega=2*pi/28,AmpSin=0,PhiSin=0)
upper<-c(Wake=Inf,Sleep=Inf,loggamma=Inf,omega=2*pi/20,AmpSin=Inf,PhiSin=2*pi)
# Fit
optimxres<-optimx(params,objfun,method=c("nlminb"),lower = lower,upper=upper)
optimxres
```


Get fit

```{r}
out<-SWDMrFit(model,params = optimxres[1,])
par(mfrow=c(2,1))
plot(out$time,out$y1,type="l",ylab="Position",xlab="Time")
plot(out$time,out$y2,type="l",ylab="Speed",xlab="Time")
```

Get some statistics for the fit

```{r}
SWDMrStats(model,out,detailed = T)
```

Plot fit and gene expression using ggplot2

```{r}
SWDMr:::StandardFittingPlot(model,optimxres[1,])
```

Compute confidence interval using empirical bootstrap

```{r}
L<-300
B <- matrix(nrow=L, ncol=length(params))

F <- matrix(nrow=L,ncol=length(out$y1))
Y <- matrix(nrow=L,ncol=length(out$y1))

pred<-SWDMrFit(model,params = optimxres[1,])
predv<-pred$y1
names(predv)<-pred$time

predval<-predv[as.character(model@Gexp[,"Time"])]


stats<-SWDMrStats(model,out,detailed = T)

model2<-model

paramsboot<-as.numeric(optimxres[,names(params)])
names(paramsboot)<-names(params)

for (l in (1:L)) {
  #if (l %%10 == 0){print(l)}
  n<-length(model2@Gexp[,1])
  model2@Gexp[,1]<-predval + sqrt(stats[["ErrorVariance"]])*rnorm(n)
  
  objfunboot<-SWDMrGetEvalFun(model2)
  optimxresboot<-optimx(paramsboot,objfunboot,method=c("nlminb"),lower = lower,upper=upper)
  
  B[l,] <- as.numeric(optimxresboot["nlminb",names(params)])
  
  outboot<-SWDMrFit(model,params = optimxresboot)
  F[l,] <- outboot$y1
  Y[l,] <- outboot$y1 + rnorm(1,0,sqrt(stats[["ErrorVariance"]]))
}

df.mc<-data.frame(Time=out$time)
df.mc[c("lwr.conf","upr.conf")] <- t(apply(F,MARGIN=2,function(x) quantile(x,c(alpha/2,1-alpha/2))))
df.mc[c("lwr.pred","upr.pred")] <- t(apply(Y,MARGIN=2,function(x) quantile(x,c(alpha/2,1-alpha/2))))
df.mc<-df.mc[df.mc$Time>0,]
V.mc <- cov(B)
V.mc

se.mc <- sqrt(diag(V.mc))
se.mc

t.stat <- paramsboot/se.mc
p.value <- 2*(1 - pt(abs(t.stat),51))
p.value

cbind(Estimate=paramsboot,SE=se.mc,Tstat=t.stat,p.value)

level <- 0.95
alpha <- 1 - level
b <- c(apply(B,MARGIN=2,function(x) quantile(x,alpha/2)),
       apply(B,MARGIN=2,function(x) quantile(x,1-alpha/2)))
ci.mc=matrix(b,ncol=2)
row.names(ci.mc) <- names(paramsboot)
colnames(ci.mc) <- c(paste0((1-level)/2*100,"%"),paste0((1+level)/2*100,"%"))
ci.mc
```

```{r}
p<-SWDMr:::StandardFittingPlot(model,optimxres[1,])
p <- p + annotate("ribbon",x=df.mc$Time, ymin=df.mc$lwr.pred, ymax=df.mc$upr.pred, alpha=0.2, fill="blue")
p <- p + annotate("ribbon",x=df.mc$Time, ymin=df.mc$lwr.conf, ymax=df.mc$upr.conf, alpha=0.5, fill="#339900")
p
```



```{r}
sessionInfo()
```

